jkube-kit/build/service/buildpacks/src/main/java/org/eclipse/jkube/kit/service/buildpacks/BuildPackCliDownloader.java

```diff
@@ -0,0 +1,177 @@
+import java.util.Collections;
+import java.util.Properties;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static java.nio.file.StandardCopyOption.COPY_ATTRIBUTES;
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+import static org.eclipse.jkube.kit.common.util.EnvUtil.findBinaryFileInUserPath;
+import static org.eclipse.jkube.kit.common.util.IoUtil.downloadArchive;
+import static org.eclipse.jkube.kit.common.util.SemanticVersionUtil.removeBuildMetadata;
+
+public class BuildPackCliDownloader {
+  private static final String JKUBE_PACK_DIR = ".jkube";
+  private static final String PACK_UNIX_CLI_NAME = "pack";
+  private static final String PACK_WINDOWS_CLI_NAME = "pack.exe";
+  private static final String PACK_DEFAULT_CLI_VERSION_PROPERTY = "version";
+  private static final String PACK_CLI_LINUX_ARTIFACT = "linux.artifact";
+  private static final String PACK_CLI_LINUX_ARM64_ARTIFACT = "linux-arm64.artifact";
+  private static final String PACK_CLI_MACOS_ARTIFACT = "macos.artifact";
+  private static final String PACK_CLI_MACOS_ARM64_ARTIFACT = "macos-arm64.artifact";
+  private static final String PACK_CLI_WINDOWS_ARTIFACT = "windows.artifact";
+
+  private final KitLogger kitLogger;
+  private final String packCliVersion;
+  private final String packCliLinuxArtifact;
+  private final String packCliLinuxArm64Artifact;
+  private final String packCliMacOsArtifact;
+  private final String packCliMacOsArm64Artifact;
+  private final String packCliWindowsArtifact;
+  private final Properties packProperties;
+
+  public BuildPackCliDownloader(KitLogger kitLogger, Properties packProperties) {
+    this.kitLogger = kitLogger;
+    this.packProperties = packProperties;
+    packCliVersion = (String) packProperties.get(PACK_DEFAULT_CLI_VERSION_PROPERTY);
+    packCliLinuxArtifact = (String) packProperties.get(PACK_CLI_LINUX_ARTIFACT);
+    packCliLinuxArm64Artifact = (String) packProperties.get(PACK_CLI_LINUX_ARM64_ARTIFACT);
+    packCliMacOsArtifact = (String) packProperties.get(PACK_CLI_MACOS_ARTIFACT);
+    packCliMacOsArm64Artifact = (String) packProperties.get(PACK_CLI_MACOS_ARM64_ARTIFACT);
+    packCliWindowsArtifact = (String) packProperties.get(PACK_CLI_WINDOWS_ARTIFACT);
+  }
+
+  public File getPackCLIIfPresentOrDownload() {
+    try {
+      File jKubeUserHomeDir = new File(getUserHome(), JKUBE_PACK_DIR);
+      FileUtil.createDirectory(jKubeUserHomeDir);
+      File pack = findPreviouslyDownloadedPackCLIPath(jKubeUserHomeDir);
+      File packInJKubeUserHome = new File(jKubeUserHomeDir, getPlatformBinaryName());
+
+      if (pack == null) {
+        File tempDownloadDirectory = FileUtil.createTempDirectory();
+        URL downloadUrl = new URL(inferApplicableDownloadArtifactUrl());
+        File downloadedPackBinary = downloadPackCli(downloadUrl, tempDownloadDirectory);
+        Files.copy(downloadedPackBinary.toPath(), packInJKubeUserHome.toPath(), REPLACE_EXISTING, COPY_ATTRIBUTES);
```

Comment:
Maybe replace the `getPlatformBinaryName` method to `resolveBinaryLocation` which returns a `Path`

Reply:
In general, this method should be something like:
- Resolve expected Path binary location ---> resolveBinaryLocation
- Check if exists and valid: packPath.exists() && isValid
- If it doesn't exist:
  - Download
- Return packPath